torch/csrc/autograd/engine.cpp

```diff
@@ -313,45 +319,33 @@ auto Engine::thread_main(
         thread_on_exception(local_graph_task, task.fn_, e);
       }
     }
-    // Notify downstream about the completion of tasks depending
-    // on both where the task was executed, and who owned the overall
-    // graph (in case of reentrant execution.)  See Note [Reentrant backwards].
-    auto base_owner = local_graph_task->owner_;
-    // Task from a non-worker thread. Easy case.
-    if (base_owner == NO_DEVICE) {
-      if (--local_graph_task->outstanding_tasks_ == 0) {
-        // Lock mutex to notify the GraphTask waiting on not_done_
-        std::lock_guard<std::mutex> lock(local_graph_task->mutex_);
-        local_graph_task->not_done_.notify_all();
-      }
-    } else {
-      // If it's a task initiated from this thread, decrease the counter, but
-      // don't do anything - loop condition will do all checks for us next.
-      if (base_owner == worker_device) {
-        --local_graph_task->outstanding_tasks_;
-        // Otherwise send a dummy function task to the owning thread just to
-        // ensure that it's not sleeping. If it has work, it might see that
-        // graph_task->outstanding_tasks_ == 0 before it gets to the task, but
-        // it's a no-op anyway.
-      } else if (base_owner != worker_device) {
-        if (--local_graph_task->outstanding_tasks_ == 0) {
-          // Synchronize outstanding_tasks_ with queue mutex
-          std::atomic_thread_fence(std::memory_order_release);
-          ready_queue_by_index(base_owner)
-              .push(NodeTask(local_graph_task, nullptr, InputBuffer(0)));
-        }
-      }
+
+    // Decrement the outstanding tasks.
+    --local_graph_task->outstanding_tasks_;
+
+    // Check if we've completed execution.
+    bool gt_completed = graph_task_completed(local_graph_task);
+    if (gt_completed) {
+      // We don't need to explicitly notify the owner thread, since
+      // 'mark_graph_task_completed' would mark the Future as completed and this
+      // would notify the owner thread that the task has been completed.
+      mark_graph_task_completed(local_graph_task);
     }
-  }
 
-  // When current_depth is 0 this worker thread is done and we need to notify
-  // the parent thread waiting on the graph_task
-  // NOTE: An edge case for this is when reentrant calls are repeatedly made in
-  // a thread which is at its maximum stack depth and they keep exiting right
-  // after. We will always switch to a new thread for each call, so, we'll keep
-  // oscillating between the two threads.
-  if (reentrant_thread && current_depth == 0) {
```

Comment:
seems like we don't handle this case in the new future way anymore. we should mark the future as completed for this condition as well to make sure parent thread know it.

Reply:
IIUC, wouldn't each invocation of `thread_main` at least enter the while loop once (since outstanding tasks would be > 0)? If so, wouldn't `mark_graph_task_completed` inside the loop notify the future appropriately? Are there any cases where this wouldn't happen?

Reply:
hmm I think you are right. If it is a re-entrant thread, then the loop will only terminate when `outstanding_tasks_ == 0`, and that will trigger the future being marked as complete, and should be ok for us to remove this.